Miflora batteries

PeteBa opened the pull request:
**Describe the change**
Adds support for miflora battery sensors. Queries the BLE Service/Characteristic as the information is not provided by advertisement.

There are four commits that cover:
1) Tweaks to existing xiaomi integration: a) extend use of typing; b) simplify handling of sensor events; c) deduping advertisements using frame count; d) detuning some logging to avoid spam
2) Extract BLE query routines from BLE integration and make available as Bluetooth support routines
3) Add polling of the battery service/characteristic on a daily basis (enabled/disabled via config)
4) Re-aligned and added tests for new functionality

The above required changing xiaomi `handleDiscovery()` to be async along the same lines as the BLE integration. Reviewing the existing code suggests this is OK but let me know any concerns. I tried not to change any of the BLE query logic in the process other than moving it from BluetoothLowEnergy to BluetoothService integration. 

I'm doing some more soak testing as BLE queries don't seem quite as stable as advertisements. So marked this as WIP but appreciate any thoughts and happy to adapt. Documentation to follow.

**Checklist**
If you changed code:
- [x] Tests run locally and pass (`npm test`)
- [x] Code has correct format (`npm run format`)

If you added a new integration:
- [ ] Documentation page added in `docs/integrations/`
- [ ] Page linked in `docs/.vuepress/config.ts` and `docs/integrations/README.md`

**Additional information**
Is this PR related to any issues? Do you have anything else to add?

Closes #638

PeteBa commented:
@mKeRix, just wanted to reach out with an observations on BLE query functionality (these appear in both this and the main branch). Basically, I am seeing quite a few occurrences of "long lock" that occurs immediately after a connection attempt fails: 

```04/08/2021, 13:13:57 - info - BluetoothLowEnergyService: Attempting app discovery for tag 75966b83f401
04/08/2021, 13:14:07 - error - BluetoothService: Failed to connect to 75:96:6b:83:f4:01: timed out
04/08/2021, 13:14:07 - warn - BluetoothLowEnergyService: Failed to discover companion app ID due to error: timed out
04/08/2021, 13:14:35 - info - BluetoothService: Detected unusually long lock on Bluetooth adapter 0, force unlocking
```
Looks like a timeout in `connectLowEnergyDevice` [here](https://github.com/mKeRix/room-assistant/blob/cb4ab598870b8db71b6c19e7527a03bd5deadfc9/src/integration-support/bluetooth/bluetooth.service.ts#L144) does not free the lock from a few lines above. This results in a lock-out period until `unlockDeadlockedAdapters` picks it up. Would it be possible to call `unlockAdapeter` in the catch ? I may be misreading this so apologies if I've missed something obvious.

PeteBa commented:
I am seeing a fairly rare error that seems to be down to the frequency that iPhones send announcements (e.g. ~five per second). This causes repeated calls to `getCompanionAppId` and exhausting the number of disconnect listeners available [here](https://github.com/mKeRix/room-assistant/blob/cb4ab598870b8db71b6c19e7527a03bd5deadfc9/src/integration-support/bluetooth/bluetooth.service.ts#L197).

```
...
06/08/2021, 00:26:43 - info - BluetoothLowEnergyService: Attempting app discovery for tag 46e6f9df5151
06/08/2021, 00:26:43 - info - BluetoothLowEnergyService: Attempting app discovery for tag 46e6f9df5151
06/08/2021, 00:26:43 - info - BluetoothLowEnergyService: Attempting app discovery for tag 46e6f9df5151
06/08/2021, 00:26:43 - info - BluetoothLowEnergyService: Attempting app discovery for tag 46e6f9df5151
06/08/2021, 00:26:44 - info - BluetoothLowEnergyService: Attempting app discovery for tag 46e6f9df5151
06/08/2021, 00:26:44 - info - BluetoothLowEnergyService: Attempting app discovery for tag 46e6f9df5151
(node:20502) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 disconnect listeners added to [Peripheral]. Use emitter.setMaxListeners() to increase limit
(Use `node --trace-warnings ...` to show where the warning was created)

I think it might be possible to defer the `promisify` call till the start of the try block a bit lower so that the listener is only established when needed. Alternatively, some logic in `discoverHandler` to limit multiple calls from the same tag could be useful.

PeteBa commented:
So I'm not sure how much of a problem this one is but when a `promiseWithTimeout` race ends then the unsettled promises are still running in the background. This happens in both xiaomi and companion app scenarios. In the particular situation below, there continues to be connection attempts well after the overall timeout has triggered and the originating `handleDiscovery` has completed. I dont fully understand the noble state machine but could this be causing the subsequent adapter resets ? One option would be having the race entrants monitor a `race_finished` flag so they cleanly terminate when losing the race.

```
07/08/2021, 14:52:47 - info - XiaomiMiService: MiFlora 17: Reading battery level
07/08/2021, 14:52:47 - info - BluetoothService:     Calling ConnectLowEnergyDevice c4:7c:8d:6b:a3:78  (queryLowEnergyDevice)
07/08/2021, 14:52:47 - info - BluetoothService:         Calling LockAdapter c4:7c:8d:6b:a3:78 (connectLowEnergyDevice)
07/08/2021, 14:52:47 - info - BluetoothService:             Locking adapter 0 (lockAdapter)
07/08/2021, 14:52:47 - info - BluetoothService:         Calling ConnectWithRetries c4:7c:8d:6b:a3:78 (connectLowEnergyDevice)
07/08/2021, 14:52:47 - info - BluetoothService:             Calling ConnectAsync at address c4:7c:8d:6b:a3:78, Tries: 5 (connectLowEnergyDeviceWithRetry)
07/08/2021, 14:52:47 - error- BluetoothService:         Exception Caught: Failed to connect to c4:7c:8d:6b:a3:78: Unknown Connection Identifier (0x2) (connectLowEnergyDevice)
07/08/2021, 14:52:47 - info - BluetoothService:         Disconnecting and removing listeners c4:7c:8d:6b:a3:78 (connectLowEnergyDevice)
07/08/2021, 14:52:47 - warn - XiaomiMiService: MiFlora 17: Error reading battery level (attempt 1 of 3): Error: Unknown Connection Identifier (0x2)

07/08/2021, 14:52:47 - info - BluetoothService:             Connection attempt returned with adapter in disconnecting - c4:7c:8d:6b:a3:78, Tries: 5 (connectLowEnergyDeviceWithRetry)
07/08/2021, 14:52:47 - info - BluetoothService:             Calling ConnectAsync at address c4:7c:8d:6b:a3:78, Tries: 4 (connectLowEnergyDeviceWithRetry)
07/08/2021, 14:52:49 - info - BluetoothService:             Connection attempt returned with adapter in disconnected - c4:7c:8d:6b:a3:78, Tries: 4 (connectLowEnergyDeviceWithRetry)
07/08/2021, 14:52:49 - info - BluetoothService:             Calling ConnectAsync at address c4:7c:8d:6b:a3:78, Tries: 3 (connectLowEnergyDeviceWithRetry)
07/08/2021, 14:53:29 - info - BluetoothService:             Connection attempt returned with adapter in disconnected - c4:7c:8d:6b:a3:78, Tries: 3 (connectLowEnergyDeviceWithRetry)
07/08/2021, 14:53:29 - info - BluetoothService:             Calling ConnectAsync at address c4:7c:8d:6b:a3:78, Tries: 2 (connectLowEnergyDeviceWithRetry)
07/08/2021, 14:54:19 - info - BluetoothService:             Connection attempt returned with adapter in disconnected - c4:7c:8d:6b:a3:78, Tries: 2 (connectLowEnergyDeviceWithRetry)
07/08/2021, 14:54:19 - info - BluetoothService:             Calling ConnectAsync at address c4:7c:8d:6b:a3:78, Tries: 1 (connectLowEnergyDeviceWithRetry)

07/08/2021, 14:55:00 - warn - BluetoothService: Did not detect any low energy advertisements in a while, resetting
07/08/2021, 14:56:03 - warn - BluetoothService: Did not detect any low energy advertisements in a while, resetting
07/08/2021, 14:56:48 - warn - BluetoothService: Did not detect any low energy advertisements in a while, resetting
```

Secondary concern is the "Unknown Connection Identifier (0x2)" error generated within hci_socket. Not sure what this error means but I have also seen the above scenario with other connection failure reasons.

PeteBa commented:
Sorry for the download above on what are somewhat tangential to monitoring miflora batteries 8) Let me know if it would be useful for me to spin up a separate PR to track any of these as happy to try and work up solutions if that would be helpful.

mKeRix commented:
Sorry for not being responsive on this PR! I actually started reviewing it last Friday, but haven't found the time to finish yet. Overall the changes looked good at least.

As for your remarks:

- The adapter unlock in the catch block was supposed to happen through the disconnect listener on the async when the code was originally written - however, as you realized this doesn't necessarily work for that block anymore since the introduction of the retry logic (great catch!). I think calling unlockAdapter in there makes total sense.
- I think both of your suggestions are good calls! The simpler thing for now would simply be moving creation of the listener. Deduplication of discovery calls would be a nice thing to have, but we can just open up an issue for tracking that and take care of it later on.
- Again, good catch! The connect promise was supposed to be terminated by the disconnect call on the peripheral, but this will of course be ignored by the retries if they aren't exceeded already. Since a lot of adapters don't support scanning and connecting at the same time this may very well be causing the adapter to not deliver any advertisements any more and therefore be reset afterwards. I can't think of a better way to work around this than a cancellation token. Might make sense to build this based on a library, such as `@esfx/async-canceltoke` ([usage example](https://github.com/rbuckton/prex/issues/7#issuecomment-747705759)). Any other or no lib is also fine by me though.
- HCI sockets and noble are fickle beasts unfortunately, I've never understood some of the errors that they generate. :( I was playing with the thought of migrating all the noble calls to the BlueZ D-Bus API in the hopes that their implementations/handling is better, but that will be quite a big change (and I'm not even sure if it makes a notable difference).

In general having a separate PR would be nicer, since stuffing them all into this one would probably end up hiding the changes in the squashed commit... but I also understand if you'd prefer to stick to this PR to avoid having too many versions of the software to work on at the same time.

PeteBa commented:
OK, so I've removed the WIP tag and think this PR is ready from my side. I have added one last commit for device firmware information given it was already available in the battery characteristic message. Appreciate that this isn't an easy PR to review given the cut/paste between files but hopefully each commit is digestible !?! Let me know any areas that could do with some supporting commentary.

I'll take a crack at the BLE stuff in another PR to keep things transparent. Will base off of this feature branch as it allows me to test with two iphones and 17 mifloras so gets a work out. I'll take a look at esfx.

mKeRix reviewed (approved):
Really sorry for letting this one sit for so long... life got in the way. The changes look good to me (really appreciate the additional effort you put into refactoring as well!). I'll take the changes for a spin on my local setup as well and will trigger a release after that.

PeteBa commented:
Thanks for this and appreciate taking the changes on-board. I think it might make sense to combine #887 when releasing this PR so that some of the issues identified abov eare mitigated. I am pretty close to finishing up 887 so I hope that make sense.

lesensei commented:
Sorry if I'm hijacking this thread. Tell if I should open a new issue instead. I was just wondering if this could serve as groundwork for polling a miband 1S (or other unencrypted mibands) for battery level (which is not advertised but can be queried with gatttool, for instance). What do you think ?
